tests/components/tado/test_service.py

```diff
@@ -0,0 +1,139 @@
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+        assert response is None
+
+
+async def test_add_meter_readings_exception(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add_meter_readings service with a None response (generic RequestExceptionwas raised)."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    with (
+        patch(
+            "homeassistant.components.tado.TadoConnector.set_meter_reading",
+            return_value=None,
+        ),
+        pytest.raises(ServiceValidationError) as exc,
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
+        assert "Could not add meter reading" in str(exc)
+
+
+async def test_add_meter_readings_invalid(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add_meter_readings service with an invalid_meter_reading response."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    fixture: str = load_fixture("tado/add_readings_invalid_meter_reading.json")
+    with (
+        patch(
+            "homeassistant.components.tado.TadoConnector.set_meter_reading",
+            return_value=json.loads(fixture),
+        ),
+        pytest.raises(ServiceValidationError) as exc,
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
+        assert "invalid new reading" in str(exc)
+
+
+async def test_add_meter_readings_duplicate(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add_meter_readings service with a duplicated_meter_reading response."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    fixture: str = load_fixture("tado/add_readings_duplicated_meter_reading.json")
+    with (
+        patch(
+            "homeassistant.components.tado.TadoConnector.set_meter_reading",
+            return_value=json.loads(fixture),
+        ),
+        pytest.raises(ServiceValidationError) as exc,
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
+        assert "reading already exists for date" in str(exc)
```

Comment:
Why not assert the entire JSON response? It's mocked so it can be checked entirely.

Reply:
imho; since it's mocked, we're testing that the mock works. Why be over extensive about it.

Reply:
Why not be complete? You've got everything already there. At least check on the error_code, whereas the description might vary.

Reply:
only the `message` is returned in the exception